File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
      
      def isPossible(capacity):
        days_needed = 1
        current_load = 0
        
        for weight in weights:
          current_load += weight
          if current_load > capacity:
            days_needed += 1
            current_load = weight
            
        return days_needed <= days
      
      
      l = max(weights)
      r = sum(weights)
      
      while l < r:
        mid = math.floor((l + r)/2)
        feasible = isPossible(mid)
        if feasible:
          r = mid
        else:
          l = mid + 1
      
      return l

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */
var shipWithinDays = function(weights, days) {
 
  function isPossible(capacity) { 
    let currentLoad = 0;
    let daysNeeded = 1;
    
    for (let weight of weights) {
      currentLoad += weight;
      if (currentLoad > capacity) {
        daysNeeded++;
        currentLoad = weight;
      } 
    }
    return daysNeeded <= days;
  }
  
  const totalLoad = weights.reduce((a,b) => a+b, 0);
  let l = Math.max(...weights);
  let r = totalLoad;  
  
  while (l < r) {
    const mid = Math.floor((l + r) / 2);
    const feasible = isPossible(mid);
    if (feasible) {
      r = mid;
    } else {
      l = mid + 1;
    }
  }
  
  return l;
};